django/dkg/cancer/json_api/endpoints.py

```diff
@@ -0,0 +1,218 @@
+        django.conf.settings.LABEL_CODES_PATH
+    )
+
+    # save predicted keywords with their associated probability
+    for index in range(label_probas.shape[0]):
+        for keyword_probability, keyword in zip(label_probas[index, :], label_names):
+            kw = RISArticleKeyword(
+                ris_article=db_articles[index],
+                keyword=keyword,
+                keyword_probability=keyword_probability,
+                annotator_name="scikit-model-1.0"
+            )
+            kw.save()
+
+    # select all of the inference articles with keywords
+    db_articles_with_keywords = all_articles_by('INFERENCE')
+    all_inference_articles = []
+    for article in db_articles_with_keywords:
+
+        entry = {
+            'article_id': article.id,
+            'title': article.title,
+            'abstract': article.abstract
+        }
+
+        entry_predictions = []
+        for keyword, keyword_probability in zip(article.ts_keywords.split("\t"), article.ts_keyword_probabilities.split("\t")):
+            distance_to_hyperplane = scipy.special.logit(float(keyword_probability))
+
+            entry_predictions.append(
+                {
+                    'keyword': keyword,
+                    'probability': keyword_probability,
+                    'distance_to_hyperplane': distance_to_hyperplane
+                }
+            )
+
+        entry['keywords'] = entry_predictions
+        all_inference_articles.append(entry)
+
+    # TODO: make active learning strategy part of POST-request
+    # active learning component for sorting
+    import cancer.active_learning.selection_strategies
+    sorted_articles = cancer.active_learning.selection_strategies.SelectionStrategies.default(
+        all_inference_articles
+    )
+
+    return JsonResponse(
+        {
+            'article_predictions': sorted_articles
+        }
+    )
+
+
+def update_model_with_feedback():
+    """
+    moves inference documents with feedback to train corpus and updates the model including that data.
+    """
+    # TODO: move data from INFERENCE to TRAIN
+
+    # rerun model selection and persist
+    cancer.model_api.model.update_model()
+
+
+def feedback(request):
+    """
+    Receives a POST-request with a single document-label feedback pair:
+    {
+        'article_id': 123,
+        'keyword': '4-harn,pall',
+        'vote': 'OK',
+        'annotator_name': 'annotator user name'
+    }
+
+    'vote' is in the set('OK', 'NOT OK').
+
+    depending on 'vote' the keyword is either added or removed from the articles keywords.
+    """
+    should_retrain = False
+    if should_retrain:
+        # TODO: since there are all keywords stored per default
+        # TODO: how to find a good strategy for label selection going from inf->train? (thresholding?)
+        update_model_with_feedback()
+
+    request_body = json.loads(request.body)
+
+    # TODO: duplicate keywords possible
+
+    if request_body['vote'] == 'OK':
```

Comment:
Especially if there's disagreement between the votes, we would need to integrate multiple feedbacks per article - but maybe they really do only incorporate votes they have discussed already and merged.